TensorFlow image_processing component

hunterjm opened the pull request:
## Description:
Adding a [TensorFlow Object Detection](https://github.com/tensorflow/models/blob/master/research/object_detection/README.md) image_processing component to run either pre-trained or custom built machine learning object detection models through TensorFlow and expose the results to Home Assistant.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7083

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Basic configuration.yaml entry
image_processing:
  - platform: tensorflow
    source:
      - entity_id: camera.local_file
    model:
      graph: /home/homeassistant/.homeassistant/frozen_inference_graph.pb
      labels: /home/homeassistant/.homeassistant/tensorflow/object_detection/data/mscoco_label_map.pbtxt
      model_dir: /home/homeassistant/.homeassistant/tensorflow

# Advanced configuration.yaml entry
image_processing:
  - platform: tensorflow
    source:
      - entity_id: camera.local_file
    file_out:
      - "/tmp/{{ camera_entity.split('.')[1] }}_latest.jpg"
      - "/tmp/{{ camera_entity.split('.')[1] }}_{{ now().strftime('%Y%m%d_%H%M%S') }}.jpg"
    model:
      graph: /home/homeassistant/.homeassistant/frozen_inference_graph.pb
      labels: /home/homeassistant/.homeassistant/tensorflow/object_detection/data/mscoco_label_map.pbtxt
      model_dir: /home/homeassistant/.homeassistant/tensorflow
      categories:
        - category: person
          area:
            # Exclude top 10% of image
            top: 0.1
            # Exclude right 15% of image
            right: 0.85
        - car
        - truck
```

## Checklist:
  - [x] The code change is tested and works locally.
    - Also tested by @arsaboo 
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable.
  - [x] New dependencies are only imported inside functions that use them.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

hunterjm committed:
initial tensorflow image_processing component

hunterjm committed:
linting fixes

robmarkcole commented:
@hunterjm great work on this. Couple of comments:

1. Re bounding box, in [Facebox](https://www.home-assistant.io/components/image_processing.facebox/) we are using `bounding_box` rather than `box`,  My plan was to roll this const into the `image_processing` component. I think it would be good to decide a convention. Also  is the `score` use here different in intent from the `confidence` used by Facebox etc? 

2. On a related point, it would be good to breakout the bounding box display as a service which all image processing components could use, I was actually about to start work on a PR to do this :-)

Cheers

robmarkcole commented:
Currently the state data will require parsing:
```
{ "dog": [ { "score": 99.44924116134644, "box": [ 0.0563804991543293, 0.18027611076831818, 0.6681657433509827, 0.6643975973129272 ] } ], "chair": [ { "score": 96.00539803504944, "box": [ 0.015583686530590057, 0, 0.35551923513412476, 0.29479289054870605 ] } ], "potted plant": [ { "score": 94.95822787284851, "box": [ 0.10585738718509674, 0.5190292596817017, 0.5562731027603149, 0.9891135692596436 ] } ], "bowl": [ { "score": 86.03321313858032, "box": [ 0.43573233485221863, 0.16289281845092773, 0.8424050807952881, 0.7881014943122864 ] } ] }
```
In Facebox I use a little trick to breakout the `keys` and format the confidence to 2dp [here](https://github.com/home-assistant/home-assistant/blob/901c4f18cb441b04db30b4f883292f1d73e4ae89/homeassistant/components/image_processing/facebox.py#L78).

robmarkcole commented:
With `file_out` configured I get the exception:
```
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/tensorflow.py", line 334, in process_image
    self._save_image(image, matches, paths)
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/image_processing/tensorflow.py", line 240, in _save_image
    if self._category_areas[category] != [0, 0, 1, 1]:
KeyError: 'dog'
```

hunterjm commented:
@robmarkcole - There is a lot here, so I will break it down by section:

> in Facebox we are using `bounding_box` rather than `box` ... is the `score` use here different in intent from the `confidence` used by Facebox etc?

Score has the same intent.  I have updated the code to reflect and keep standard.  If we want to standardize bounding box, we can as well.  TensorFlow returns the `bounding_box` as `[top, left, bottom, right]`.  Is that how Facebox defines it as well?

> It would be good to breakout the bounding box display as a service which all image processing components could use.

I agree, but I think we can do that (potentially together) as a refactor PR instead of making this one larger in scope.

> I get the log error `Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA`

Thanks for the fix.  It's a harmless warning, but can see how it could confuse the standard user.

> In Facebox I use a little trick to breakout the keys and format the confidence to 2dp here.

I looked at what you were doing in Facebox, and drew some inspiration from there.  This one is a little peculiar since each category could have multiple detections.  With Facebox, you are breaking out a known matched face, and there can only be one of those.  In this scenario, there can be more than one `person` or `car` detected.  I'm happy to revisit the data structure, but I went with what I thought was most intuitive at the time.

> With `file_out` configured I get [an] exception

Great catch!  I forgot to test `file_out` without filtered `categories`.  Fixed in https://github.com/home-assistant/home-assistant/pull/17795/commits/8bb9cd37e276efefd0d21bc15b4a72cce355dbb7

> If I use the camera entity_id in the template I get an error: Error rendering template: UndefinedError: 'camera' is undefined.

You are probably using `camera.entity_id`, which isn't known by the template.  The [documentation](https://deploy-preview-7083--home-assistant-docs.netlify.com/components/image_processing.tensorflow/#file_out) lets people know to use `camera_entity` which is the entity string (i.e. `camera.backyard`).  I'm happy to change if you think there will be confusion.

robmarkcole commented:
@hunterjm I like the convention for bounding box you propose (`[top, left, bottom, right]`), and suggest we stick with that and refactor the other components later. We can then just provide [a link](https://www.tensorflow.org/api_docs/python/tf/image/draw_bounding_boxes) to the convention for future reference. 

Re displaying the returned data, you currently display:
```python
{'dog': [{'confidence': 99.45,
   'bounding_box': [0.0563804991543293,
    0.18027611076831818,
    0.6681657433509827,
    0.6643975973129272]}],
 'chair': [{'confidence': 96.01,
   'bounding_box': [0.015583686530590057,
    0,
    0.35551923513412476,
    0.29479289054870605]}],
```
What I propose is that in addition to this fill info, we provide a summary:
```python
{'dog': : 99.45,
 'chair': 96.01,
```
This is [the approach in Facebox](https://github.com/robmarkcole/HASS-Machinebox-Facebox/blob/master/usage.png), and makes it easy to see whats been identified from the HA front end:

arsaboo commented:
@robmarkcole We should try to be close to Facebox, but it is not a perfect match as there can be more than one object detected. We can add a summary, but I feel that it is redundant. We can use `confidence: 75` in the configuration to exclude matches below a certain threshold (@hunterjm we should add that as an optional configuration setting). 

For `file_out`, I just tried the following and it works fine for me:

```yaml
    file_out:
      - "/home/homeassistant/.homeassistant/downloads/camera/test_{{ now().strftime('%Y%m%d_%H%M%S') }}.jpg"
```

robmarkcole commented:
[DEFAULT_CONFIDENCE](https://github.com/home-assistant/home-assistant/blob/92e9c2aa724bee8e7a29cc72b0ab3253b451fea7/homeassistant/components/image_processing/__init__.py#L51) is already part of the platform

hunterjm commented:
@robmarkcole - I hear you on the summary.  I still don't think having `key = category` and `value = confidence` is the right take though.  In the example you provided, there were 2 cups detected.  What about quantity detected, like this (taken from your example [here](https://imgur.com/a/ysLbYId)):

```
{
  "cup": 2,
  "fork": 1,
  "dining table": 1,
  "pizza": 1
}
```

robmarkcole commented:
@hunterjm yes I see what you mean by 'multiple objects being detected' now, your suggestion looks great then.

Next people will want to create automations based on detections, the way to do this is to add `events`, like [here](https://github.com/home-assistant/home-assistant/blob/9cd8a86eb4f9fec8f32105aa0bac26d70a02c059/homeassistant/components/image_processing/__init__.py#L216). Possibly beyond the scope of this PR.

hunterjm committed:
make displayed attribute a summary of objects

hunterjm committed:
fix missed merge conflict and add warning supression back in for CPU type

robmarkcole commented:
@hunterjm did you decide to drop `os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'` then?
Next step is to add unit tests, and hit 95% coverage. You should be able to steal from [existing](https://github.com/home-assistant/home-assistant/tree/dev/tests/components/image_processing) tests.

hunterjm commented:
@Robin - I did not mean to. It got dropped in the rebase.

balloob commented on Dockerfile:
```diff
@@ -25,10 +28,11 @@ RUN virtualization/Docker/setup_docker_prereqs
 # See PR #8103 for more info.
 RUN pip3 install --no-cache-dir -r requirements_all.txt && \
-    pip3 install --no-cache-dir mysqlclient psycopg2 uvloop cchardet cython
+    pip3 install --no-cache-dir mysqlclient psycopg2 uvloop cchardet cython opencv-python-headless
```
I'm always a bit wary if we need to install things for every single user of Home Assistant. Why can't we install this on demand?

balloob replied:
It looks like it will also work without opencv. We should remove this.

hunterjm replied:
The thought here was to make it as easy as possible for those using the official docker image.  OpenCV is pre-built in the Hass.io images [here](https://github.com/home-assistant/hassio-homeassistant/blob/master/base/Dockerfile#L36-L57), though from source.  The pip package makes it easier to install.

I was aiming for consistency.

balloob commented on homeassistant/components/image_processing/tensorflow.py:
```diff
@@ -0,0 +1,336 @@
+CONF_BOTTOM = 'bottom'
+CONF_RIGHT = 'right'
+
+DEFAULT_MODEL_DIR = "/usr/src/app/tensorflow"
```
We should default model dir to our config dir?

hunterjm replied:
This is the default hassio config dir.  Is there a helper function available to retrieve the config dir?  If so, I can use that.

balloob commented on homeassistant/components/image_processing/tensorflow.py:
```diff
@@ -0,0 +1,336 @@
+    model_dir = model_config.get(CONF_MODEL_DIR)
+
+    # append custom model path to sys.path
+    sys.path.append(model_dir)
```
Ehhhrrrrr

We should be able to specify the paths to look to tensorflow?

hunterjm replied:
There are some requirements that can not be installed via pip.  The object detection portion of the library is not included in the standard TensorFlow wheel.  This is to include the protobuf models and util functions created as a part of the tensorflow install shell script.

balloob commented on virtualization/Docker/setup_docker_prereqs:
```diff
@@ -63,8 +64,13 @@ if [ "$INSTALL_SSOCR" == "yes" ]; then
   virtualization/Docker/scripts/ssocr
 fi
 
+<<<<<<< HEAD
```
merge conflict??

michaelarnauts commented:
These binaries/tensorflow should really be in a different container if you want to use them. Home Assistant should just contain the glue to interface with it, but not the actual code.

michaelarnauts commented on Dockerfile:
```diff
@@ -13,6 +13,9 @@ LABEL maintainer="Paulus Schoutsen <Paulus@PaulusSchoutsen.nl>"
 #ENV INSTALL_SSOCR no
 #ENV INSTALL_DLIB no
 #ENV INSTALL_IPERF3 no
+#ENV INSTALL_TENSORFLOW no
```
I really don't like that we include tensorflow in the image. You should work an a solution that uses https://hub.docker.com/r/tensorflow/tensorflow/ for example.

hunterjm replied:
The only other solution would be to use a tensorflow serving image and have this component communicate with it via gRPC.  The other option is to include TensorFlow via `REQUIREMENTS` to have it auto load, and move the install portion of the object detection models to `setup_platform`.

The tradeoffs here is we would need to default to the least common denominator - the CPU.  That means users with a CUDA enabled GPU wouldn't be able to benefit from increased inference speed.  There also isn't an official wheel for alpine linux (which Hass.io uses).  The plan was to update that repo eventually to have TensorFlow pre-built into that image as well.

homeassistant commented:
Hi @rv-jhunter, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#17795).

Once you do that we will be able to review and accept this pull request.

Thanks!

hunterjm committed:
restructure tensorflow component to install on the fly, remove from Docker

hunterjm committed:
add both matches and summary as attributes

balloob commented on homeassistant/components/image_processing/tensorflow.py:
```diff
@@ -0,0 +1,346 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the TensorFlow image processing platform."""
+    import sys
```
there are built-in modules, can be a the top of the file

balloob commented on homeassistant/components/image_processing/tensorflow.py:
```diff
@@ -0,0 +1,346 @@
+        return self._total_matches
+
+    @property
+    def state_attributes(self):
```
Should be `device_state_attributes`, `state_attributes` is reserved for entity base classes.

balloob commented on homeassistant/components/image_processing/tensorflow.py:
```diff
@@ -0,0 +1,346 @@
+        img_width, img_height = img.size
+        draw = ImageDraw.Draw(img)
+
+        def draw_box(draw, box, img_width,
```
Any reason this needs to defined inline and not outside of the class?

hunterjm replied:
Nope, resolved

hunterjm committed:
address review comments

hunterjm committed:
do not use deps folder as default, as it should only be managed by HA.  Update to have tensorflow in root config directory

loxK commented:
It is great to see this added, well done. But it won't replace my setup for security. 

Here is what would be awesome:
- first do a simple (cpu friendly) motion detection
- if motion is detected, try a person detection (tensorflow)
- if person (or other object) is detected wait x configurable seconds before doing any tensorflow object detection
- add option to stop tensorflow processing when at least x (configurable) person is detected

I have motion + a custom script (PHP) that does that, would love to replace it with home assistant.

MartinHjelmare commented:
Please open an issue if you suspect a bug.

If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion or bug reports. If you've found a bug it's ok to make a review with inline comments and link to an issue that reports the bug.

Thanks!
